File: typescript/873.length-of-longest-fibonacci-subsequence.ts
/*
 * @lc app=leetcode id=873 lang=typescript
 *
 * [873] Length of Longest Fibonacci Subsequence
 */

// @lc code=start
function lenLongestFibSubseq(arr: number[]): number {
    const numSet = new Set(arr);
    let maxLen = 0;
    const n = arr.length;

    // Try all possible first two numbers of the sequence
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            // Start with the first two numbers
            let prev = arr[j];
            let curr = arr[i] + arr[j];
            let currLen = 2;

            // Keep finding the next Fibonacci-like number
            while (numSet.has(curr)) {
                const temp = curr;
                curr = curr + prev;
                prev = temp;
                currLen++;
                maxLen = Math.max(maxLen, currLen);
            }
        }
    }
    return maxLen;
};
// @lc code=end

File: python/873.length-of-longest-fibonacci-subsequence.py
#
# @lc app=leetcode id=873 lang=python3
#
# [873] Length of Longest Fibonacci Subsequence
#

# @lc code=start
from typing import List


class Solution:
    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        # Store array elements in set for O(1) lookup
        num_set = set(arr)
        max_len = 0
        n = len(arr)

        # Try all possible first two numbers of sequence
        for start in range(n):
            for next in range(start + 1, n):
                # Start with first two numbers
                prev = arr[next]
                curr = arr[start] + arr[next]
                curr_len = 2

                # Keep finding next Fibonacci number
                while curr in num_set:
                    prev, curr = curr, curr + prev
                    curr_len += 1
                    max_len = max(max_len, curr_len)

        return max_len  


# @lc code=end